### PR TITLE
[SPARK-46654][SQL][PYTHON] Make `to_csv` explicitly indicate that it does not support some types of data

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to make `to_csv` `explicitly` indicate that it does not support `some types of data`, includes:
- Struct
- Array
- Map
- Variant
- Binary

### Why are the changes needed?
- Because the `CSV specification` does not have standards for these `types`, and cannot be read back through `from_csv`.

- For this case
  Before:
  image

  After:
  image

### Does this PR introduce _any_ user-facing change?
Yes, `to_csv` will explicitly indicate that it does not support some types of data.

### How was this patch tested?
- Add & Update UT.
- Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by panbingkun** [comment]
cc @cloud-fan @LuciferYang

**Comment by LuciferYang** [comment]
@panbingkun Let's re-enable the doctest

https://github.com/apache/spark/blob/514ecc6fc183d7222b9dc299af4df328c71966d1/python/pyspark/sql/functions/builtin.py#L15053-L15058

and remove the TODO

https://github.com/apache/spark/blob/514ecc6fc183d7222b9dc299af4df328c71966d1/python/pyspark/sql/functions/builtin.py#L15010-L15013

also cc @HyukjinKwon

**Comment by panbingkun** [comment]
> @panbingkun Let's re-enable the doctest
> 
> https://github.com/apache/spark/blob/514ecc6fc183d7222b9dc299af4df328c71966d1/python/pyspark/sql/functions/builtin.py#L15053-L15058
> 
> and remove the TODO
> 
> https://github.com/apache/spark/blob/514ecc6fc183d7222b9dc299af4df328c71966d1/python/pyspark/sql/functions/builtin.py#L15010-L15013
> 
> also cc @HyukjinKwon

Okay.

**Comment by panbingkun** [comment]
> > @panbingkun Let's re-enable the doctest
> > https://github.com/apache/spark/blob/514ecc6fc183d7222b9dc299af4df328c71966d1/python/pyspark/sql/functions/builtin.py#L15053-L15058
> > 
> > and remove the TODO
> > https://github.com/apache/spark/blob/514ecc6fc183d7222b9dc299af4df328c71966d1/python/pyspark/sql/functions/builtin.py#L15010-L15013
> > 
> > also cc @HyukjinKwon
> 
> Okay.

Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala [hunk 1]
```diff
@@ -89,8 +90,11 @@ class UnivocityGenerator(
     case udt: UserDefinedType[_] => makeConverter(udt.sqlType)
 
     case dt: DataType =>
-      (row: InternalRow, ordinal: Int) =>
-        row.get(ordinal, dt).toString
+      (row: InternalRow, ordinal: Int) => {
+        val valueUTF8String = ToPrettyString(Literal(row.get(ordinal, dt), dt),
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:94]: Does `ToPrettyString` prints nested fields in the same way as `UnivocityGenerator`. For instance, does it follow to datetime patterns specified for the CSV datasource?

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala:94]: Let me think again, maybe we need to rewrite the logic similar to `ToPrettyString ` in `UnivocityGenerator ` based on the characteristics of CSV.

Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala]: I changed the approach and enabled `UnivocityGenerator ` to support complex type (`ArrayType`, `MapType`, `StructType`) outputs.

**Review by MaxGekk — COMMENTED** [review]
Can the `from_csv` function parse the output of `to_csv()` back? Could you add tests or modify existing tests, please.

**Review by HyukjinKwon — COMMENTED** [review]
1. Does CSV source support this? I think it's disallowed
2. If we want to support this it should better be able to read it back. Problem is that CSV spec doesn't have this.

Should probably just disallow this case

**Comment by LuciferYang** [comment]
I think the current issue is that the original PySpark can display complex struct types, 

```
bin/pyspark
Python 3.11.1 (v3.11.1:a7a450f84a, Dec  6 2022, 15:24:06) [Clang 13.0.0 (clang-1300.0.29.30)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
24/02/04 14:14:41 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
24/02/04 14:14:41 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /__ / .__/\_,_/_/ /_/\_\   version 4.0.0-SNAPSHOT
      /_/

Using Python version 3.11.1 (v3.11.1:a7a450f84a, Dec  6 2022 15:24:06)
Spark context Web UI available at http://localhost:4041
Spark context available as 'sc' (master = local[*], app id = local-1707027281550).
SparkSession available as 'spark'.
>>> from pyspark.sql import Row, functions as sf
>>> data = [(1, Row(age=2, name='Alice', scores=[100, 200, 300]))]
>>> df = spark.createDataFrame(data, ("key", "value"))
>>> df.select(sf.to_csv(df.value)).show(truncate=False)
+-----------------------+                                                       
|to_csv(value)          |
+-----------------------+
|2,Alice,"[100,200,300]"|
+-----------------------+
```

but PyConnect and Scala code cannot. Ultimately, we should make them consistent. Should we consider removing the current capability of PySpark? However, this might be a breaking change.

**Comment by panbingkun** [comment]
> Can the `from_csv` function parse the output of `to_csv()` back? Could you add tests or modify existing tests, please.

For this case, currently the `from_csv` function cann't parse the output of `to_csv()` back.
But I think after that, we can improve `from_csv so that it can complete this function.

**Comment by panbingkun** [comment]
> 1. Does CSV source support this? I think it's disallowed
> 2. If we want to support this it should better be able to read it back. Problem is that CSV spec doesn't have this.
> 
> Should probably just disallow this case

I have privately discussed this issue with @LuciferYang , and in response to this situation, we actually have two options:
- In `to_csv` When encountering `complex types` in CSV, such as `ArrayType`, `MapType`, `StructType`, we can prompt an error indicating that it is not supported.
**However, in this case, in the `df. show()`, it can actually display complex types, which looks weird.**

- Another way is for us to support it. Indeed, CSV spec doesn't have this.

**Comment by HyukjinKwon** [comment]
I think we should simply disallow this. We could have a legacy conf, but it doesn't much make sense. What does it return for `ArrayType` and `MapType`?

**Comment by panbingkun** [comment]
> I think we should simply disallow this. We could have a legacy conf, but it doesn't much make sense. What does it return for `ArrayType` and `MapType`?

Okay, I have overloaded `checkInputDataTypes` for `to_csv`(`StructsToCsv`) and executed type checking on it. Let's `explicitly` indicate that it does not support `complex type data`.

File: python/pyspark/sql/functions/builtin.py [hunk 2]
```diff
@@ -15534,19 +15532,7 @@ def to_csv(col: "ColumnOrName", options: Optional[Dict[str, str]] = None) -> Col
     |      2,Alice|
     +-------------+
 
-    Example 2: Converting a complex StructType to a CSV string
```
Comment (panbingkun) [inline python/pyspark/sql/functions/builtin.py:15537]: Considering that `to_csv` no longer supports `complex type data`, we will remove this example.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 3]
```diff
@@ -260,16 +263,41 @@ case class StructsToCsv(
...
+    case _ => false
+  }
+
   @transient
   lazy val writer = new CharArrayWriter()
 
   @transient
-  lazy val inputSchema: StructType = child.dataType match {
-    case st: StructType => st
-    case other => throw new SparkIllegalArgumentException(
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:269]: Let's `merge` type checking into `one place`.

**Comment by LuciferYang** [comment]
> I think we should simply disallow this. We could have a legacy conf, but it doesn't much make sense. What does it return for `ArrayType` and `MapType`?

Should we add the legacy conf? It's hard to confirm whether any users have relied on this erroneous behavior, and directly disallowing it may bring migration costs.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 4]
```diff
@@ -260,16 +263,43 @@ case class StructsToCsv(
...
+          "inputSql" -> toSQLExpr(child),
+          "inputType" -> toSQLType(child.dataType)
+        )
+      )
+    }
+  }
+
+  @tailrec
+  private def supportDataType(dataType: DataType): Boolean = dataType match {
+    case _: VariantType => false
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:285]: Just to keep this tidy, remove the blank lines and use the `|` syntax to handle multiple cases with the same output together

**Comment by xinrong-meng** [comment]
Would you please add `[PYTHON]` to title considering it has user-facing python change? Thanks!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala (hunk 4 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala]: Done.

**Comment by panbingkun** [comment]
> Would you please add `[PYTHON]` to title considering it has user-facing python change? Thanks!

Sure, done. 😄

**Comment by LuciferYang** [comment]
Is this considered a bug fix? Do we need to backport it to branch-3.5/3.4

**Comment by panbingkun** [comment]
> Is this considered a bug fix? Do we need to backport it to branch-3.5/3.4

Yea, I think we need to backport it to `branch-3.5/3.4`.

**Comment by panbingkun** [comment]
friendly ping @HyukjinKwon,
When you are not busy, can you please continue to help review this PR?

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala [hunk 5]
```diff
@@ -294,10 +294,19 @@ class CsvFunctionsSuite extends QueryTest with SharedSparkSession {
   }
 
   test("to_csv with option (nullValue)") {
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:296]: Could you explain, please, why did you change the test?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 6]
```diff
@@ -260,16 +263,36 @@ case class StructsToCsv(
...
+          "requiredType" -> toSQLType(StructType),
+          "inputSql" -> toSQLExpr(child),
+          "inputType" -> toSQLType(child.dataType)
+        )
+      )
+    }
+  }
+
+  @tailrec
+  private def supportDataType(dataType: DataType): Boolean = dataType match {
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:284]:
```suggestion
  private def isSupportedDataType(dataType: DataType): Boolean = dataType match {
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 7]
```diff
@@ -260,16 +263,36 @@ case class StructsToCsv(
...
+    case _: VariantType | BinaryType => false
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:285]: Could you clarify this? Why don't you support the types?

File: python/pyspark/sql/functions/builtin.py [hunk 8]
```diff
@@ -15534,19 +15532,7 @@ def to_csv(col: "ColumnOrName", options: Optional[Dict[str, str]] = None) -> Col
...
-
-    >>> from pyspark.sql import Row, functions as sf
-    >>> data = [(1, Row(age=2, name='Alice', scores=[100, 200, 300]))]
-    >>> df = spark.createDataFrame(data, ("key", "value"))
-    >>> df.select(sf.to_csv(df.value)).show(truncate=False) # doctest: +SKIP
-    +-----------------------+
-    |to_csv(value)          |
-    +-----------------------+
-    |2,Alice,"[100,200,300]"|
```
Comment (MaxGekk) [inline python/pyspark/sql/functions/builtin.py:15546]: This is a breaking change in fact. We need to update the SQL migration guide at least.

Comment (LuciferYang) [reply python/pyspark/sql/functions/builtin.py:15546]: +1

Comment (panbingkun) [reply python/pyspark/sql/functions/builtin.py:15546]: Let me update.

Comment (panbingkun) [reply python/pyspark/sql/functions/builtin.py:15546]: I have updated the document `sql-migration-guide.md`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala (hunk 7 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:282]: Because `these types` cannot be `read back` through `from_csv`,
I have added related tests in UT `CsvFunctionsSuite`.

File: sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala (hunk 5 above)
Comment (panbingkun) [reply sql/core/src/test/scala/org/apache/spark/sql/CsvFunctionsSuite.scala:298]: Because `NullType` cannot be read back through `from_csv`.
In the new logic, it will throw `DATATYPE_MISMATCH.UNSUPPORTED_INPUT_TYPE` exception.
And as far as I understand, this UT is mainly to test whether `option (nullValue)` is effective,
The test `target` of changed UT are consistent with this.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala [hunk 9]
```diff
@@ -260,16 +263,33 @@ case class StructsToCsv(
       child = child,
       timeZoneId = None)
 
+  override def checkInputDataTypes(): TypeCheckResult = {
+    child.dataType match {
+      case schema: StructType if schema.map(_.dataType).forall(
+        dt => isSupportedDataType(dt)) => TypeCheckSuccess
+      case _ => DataTypeMismatch(
+        errorSubClass = "UNSUPPORTED_INPUT_TYPE",
```
Comment (panbingkun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/csvExpressions.scala:271]: Using `UNSUPPORTED_INPUT_TYPE` instead of `UNEXPECTED_INPUT_TYPE` here will make the prompt information more `clear` to the user, for example:

```
    val rows = new java.util.ArrayList[Row]()
    rows.add(Row(1L, Row(2L, "Alice",
      Map("math" -> 100L, "english" -> 200L, "science" -> null))))

    val valueSchema = StructType(Seq(
      StructField("age", LongType),
      StructField("name", StringType),
      StructField("scores", MapType(StringType, LongType))))
    val schema = StructType(Seq(
      StructField("key", LongType),
      StructField("value", valueSchema)))

    val df = spark.createDataFrame(rows, schema)

    df.select(to_csv($"value")).collect()
```

Before(Last review):
Users may feel confused about prompt: `requires the "STRUCT" type, however "value" has the type "STRUCT...`
```
[DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE] Cannot resolve "to_csv(value)" due to data type mismatch: The 1 parameter requires the "STRUCT" type, however "value" has the type "STRUCT<age: BIGINT, name: STRING, scores: MAP<STRING, BIGINT>>". SQLSTATE: 42K09;
```

After(Right now):
```
[DATATYPE_MISMATCH.UNSUPPORTED_INPUT_TYPE] Cannot resolve "to_csv(value)" due to data type mismatch: The input of `to_csv` can't be "STRUCT<age: BIGINT, name: STRING, scores: MAP<STRING, BIGINT>>" type data. SQLSTATE: 42K09;
```

File: python/pyspark/sql/functions/builtin.py (hunk 8 above)
Comment (MaxGekk) [reply python/pyspark/sql/functions/builtin.py:15546]: > I think we should simply disallow this. We could have a legacy conf, but it doesn't much make sense. What does it return for ArrayType and MapType?

@HyukjinKwon See what it does return for `ArrayType` in the example. I doubt that we should simply disallow it w/o any legacy config at least.

Comment (panbingkun) [reply python/pyspark/sql/functions/builtin.py:15546]: In `this case` of Python, it is because the data type is ArrayData (specifically implemented as `GenericArrayData`),
This case just goes into the following logic:

https://github.com/apache/spark/blob/09bdf2a9334f210ece4c23d0f3324f81113330d0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityGenerator.scala#L91-L93

https://github.com/apache/spark/blob/09bdf2a9334f210ece4c23d0f3324f81113330d0/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala#L85


And it just so happens that `GenericArrayData` overrides the `toString` method, so it appears to display the result `correctly`, perhaps it's just a `coincidence`?

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @panbingkun and @HyukjinKwon @LuciferYang @srowen for review.

**Comment by cloud-fan** [comment]
why do we remove a working feature? what's wrong with `to_csv` generating non-standard but pretty strings for these values?

**Comment by MaxGekk** [comment]
> generating non-standard but pretty strings

@cloud-fan It prints "address" of objects, not values. Do you have an example which you worry about?

**Comment by MaxGekk** [comment]
@panbingkun I forgot to ask you. Does `to_json` have the same issue?

**Comment by HyukjinKwon** [comment]
`to_json` should be fine because it has nested type representation

**Comment by cloud-fan** [comment]
So we make a breaking change simply because we don't like the object address in the string? can we fix it?

**Comment by panbingkun** [comment]
> So we make a breaking change simply because we don't like the object address in the string? can we fix it?

These types cannot be read back through 'from_csv' after generating non-standard but pretty strings through 'to_csv'.
Do we need to add configuration to bring it back?

**Comment by panbingkun** [comment]
> `to_json` should be fine because it has nested type representation

Yeah, there is no standard for nested types in `csv` format, but the `json` format has a standard.

**Comment by cloud-fan** [comment]
You never know how people will use Spark and I'm not sure why it's necessary to ban this use case. Having inconsistency between scala and python is bad. Can we fix this problem by defining a pretty string format for struct/array/map in the csv writer?

**Comment by MaxGekk** [comment]
> You never know how people will use Spark and I'm not sure why it's necessary to ban this use case. 

I believe it is not too bad to ban such output in any case (this is even insecure). @cloud-fan Do you see any use-case for that?

Array:
```scala
    val valueSchema = StructType(Seq(StructField("age", LongType), StructField("name", StringType), StructField("scores", ArrayType(LongType))))
    val schema = StructType(Seq(StructField("key", LongType), StructField("value", valueSchema)))
    val df = spark.createDataFrame(rows, schema)
    df.select(to_csv($"value")).show(false)


+--------------------------------------------------------------------------+
|to_csv(value)                                                             |
+--------------------------------------------------------------------------+
|2,Alice,org.apache.spark.sql.catalyst.expressions.UnsafeArrayData@e7bb2e54|
+--------------------------------------------------------------------------+
```

Map:
```scala
+------------------------------------------------------------------------+
|to_csv(value)                                                           |
+------------------------------------------------------------------------+
|2,Alice,org.apache.spark.sql.catalyst.expressions.UnsafeMapData@65ab9790|
+------------------------------------------------------------------------+
```

Struct:
```scala
    val rows = new java.util.ArrayList[Row]()
    rows.add(Row(1L, Row(2L, "Alice", Row(100L, 200L, null))))
    val valueSchema = StructType(Seq(StructField("age", LongType), StructField("name", StringType), StructField("scores", StructType(Seq(StructField("id1", LongType), StructField("id2", LongType), StructField("id3", LongType))))))
    val schema = StructType(Seq(StructField("key", LongType), StructField("value", valueSchema)))
    val df = spark.createDataFrame(rows, schema)

+---------------------+
|to_csv(value)        |
+---------------------+
|2,Alice,"[4,64,c8,0]"|
+---------------------+
```

**Comment by panbingkun** [comment]
- Why is the result displayed through `to_csv` inconsistency in Scala and Python for this case?
  Because this case is on the `python side`, it ultimately uses `GenericArrayData`, which happens to implement the method `toString`, so `to_csv` displays readable text.
https://github.com/apache/spark/blob/11247d804cd370aaeb88736a706c587e7f5c83b3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/GenericArrayData.scala#L85

However, on the `scala side`, it ultimately uses `UnsafeArrayData`.  `Unfortunately`, it does not implement the method `toString` (using the default `Object.toString` method), so the final `to_csv` displays `the address of the object`.

- In the implementation process of this PR, it can display `non-standard but pretty strings`,  as follows:
  https://github.com/apache/spark/pull/44665/commits/9695e975f3299556e7c268918ecd51be7a03c157
  image
  The `disadvantage` of this is that it `cannot` be `read back` through `from_csv` `at present`.
  If the final result of the discussion is acceptable, it should be easy to bring back this feature.

- Another possible compromise solution is to add a configuration (defaultly, it does `not` support displaying data of type [Array, Map, Struct ...] as `non-standard but pretty strings` through `to_csv`). If the user sets this configuration to be enabled, restore the original behavior?

**Comment by cloud-fan** [comment]
I'd prefer to fix the inconsistency. Can we use `ToPrettyString.eval` to generate pretty strings for these types in `to_csv`?

**Comment by panbingkun** [comment]
Use `ToPrettyString.eval` to generate pretty strings for these types in `to_csv`, I have also tried it in this PR. Below:
https://github.com/apache/spark/pull/44665/commits/22a7afb6951b9274cc428f71e675df233de74a8a
image
(PS: Maybe some details about `using ToPrettyString.eval` still need to be confirmed.)

If the final discussion result is to `fix their consistency`, this approach should also be possible.
WDYT @MaxGekk @HyukjinKwon @LuciferYang @xinrong-meng @srowen ?

**Comment by LuciferYang** [comment]
I don't oppose this resolution, but if `to_csv` can handle complex data structures, what would our subsequent plans be? Would `from_csv` also need to support complex data structures to make these paired functions more self-consistent? In the long term, do we need to enhance the read and write capabilities for CSV data sources to support complex data structures?

**Comment by cloud-fan** [comment]
It's ok to not support round-trip. array/struct/map is not CSV standard either. I think the principle is to avoid breaking changes if possible. Since `to_csv` already supports array/struct/map, let's keep it but fix the inconsistency issue.

**Comment by cloud-fan** [comment]
If no objections, I'll revert this tomorrow. We can have a new PR to add pretty string for array/struct/map in `to_csv`.

**Comment by panbingkun** [comment]
> If no objections, I'll revert this tomorrow. We can have a new PR to add pretty string for array/struct/map in `to_csv`.

I will submit a new PR to `add pretty string for array/struct/map` in to_csv today.

**Comment by panbingkun** [comment]
@cloud-fan 
The initial version has been submitted here: https://github.com/apache/spark/pull/45657
I am still improving it, and when it is ready, I will ask for your help in reviewing it. 
Thank you.
